lite/provider_test.go

```diff
@@ -1,149 +1,140 @@
 
-	"github.com/tendermint/tendermint/lite"
-	liteErr "github.com/tendermint/tendermint/lite/errors"
+	lerr "github.com/tendermint/tendermint/lite/errors"
+	"github.com/tendermint/tendermint/types"
+	dbm "github.com/tendermint/tendermint/libs/db"
+	log "github.com/tendermint/tendermint/libs/log"
 )
 
-// missingProvider doesn't store anything, always a miss
-// Designed as a mock for testing
+// missingProvider doesn't store anything, always a miss.
+// Designed as a mock for testing.
 type missingProvider struct{}
 
 // NewMissingProvider returns a provider which does not store anything and always misses.
-func NewMissingProvider() lite.Provider {
+func NewMissingProvider() PersistentProvider {
 	return missingProvider{}
 }
 
-func (missingProvider) StoreCommit(lite.FullCommit) error { return nil }
-func (missingProvider) GetByHeight(int64) (lite.FullCommit, error) {
-	return lite.FullCommit{}, liteErr.ErrCommitNotFound()
+func (missingProvider) SaveFullCommit(FullCommit) error { return nil }
+func (missingProvider) LatestFullCommit(chainID string, minHeight, maxHeight int64) (FullCommit, error) {
+	return FullCommit{}, lerr.ErrCommitNotFound()
 }
-func (missingProvider) GetByHash([]byte) (lite.FullCommit, error) {
-	return lite.FullCommit{}, liteErr.ErrCommitNotFound()
-}
-func (missingProvider) LatestCommit() (lite.FullCommit, error) {
-	return lite.FullCommit{}, liteErr.ErrCommitNotFound()
+func (missingProvider) ValidatorSet(chainID string, height int64) (*types.ValidatorSet, error) {
+	return nil, errors.New("missing validator set")
 }
+func (missingProvider) SetLogger(_ log.Logger) {}
 
 func TestMemProvider(t *testing.T) {
-	p := lite.NewMemStoreProvider()
+	p := NewDBProvider("mem", dbm.NewMemDB())
 	checkProvider(t, p, "test-mem", "empty")
 }
 
-func TestCacheProvider(t *testing.T) {
-	p := lite.NewCacheProvider(
+func TestMultiProvider(t *testing.T) {
+	p := NewMultiProvider(
 		NewMissingProvider(),
-		lite.NewMemStoreProvider(),
+		NewDBProvider("mem", dbm.NewMemDB()),
 		NewMissingProvider(),
 	)
 	checkProvider(t, p, "test-cache", "kjfhekfhkewhgit")
 }
 
-func checkProvider(t *testing.T, p lite.Provider, chainID, app string) {
+func checkProvider(t *testing.T, p PersistentProvider, chainID, app string) {
 	assert, require := assert.New(t), require.New(t)
 	appHash := []byte(app)
-	keys := lite.GenValKeys(5)
+	keys := genPrivKeys(5)
 	count := 10
 
-	// make a bunch of commits...
-	commits := make([]lite.FullCommit, count)
+	// Make a bunch of full commits.
+	fcz := make([]FullCommit, count)
 	for i := 0; i < count; i++ {
-		// two commits for each validator, to check how we handle dups
-		// (10, 0), (10, 1), (10, 1), (10, 2), (10, 2), ...
 		vals := keys.ToValidators(10, int64(count/2))
 		h := int64(20 + 10*i)
-		commits[i] = keys.GenFullCommit(chainID, h, nil, vals, appHash, []byte("params"), []byte("results"), 0, 5)
+		fcz[i] = keys.GenFullCommit(chainID, h, nil, vals, vals, appHash, []byte("params"), []byte("results"), 0, 5)
```

Comment:
Very hard to read. Maybe use Builder pattern.

Reply:
I agree, but punting as it's arguably out of scope, basically priority reasons.